dockertest/environment.py

```diff
@@ -79,3 +82,197 @@ def selinux_is_enforcing():
+                                    " incompatible with %s"
+                                    % (key, value, self.nvra_type))
+
+    @classmethod
+    def _store(cls, name):
+        cmd = list(cls.rpmcmd) + [name]
+        xcept_fmt = ("%s did not return expected line-delimited"
+                     " items of hash-delimited fields as"
+                     " output: %s")
+        output = cls.execute(cmd)
+        for value in output.strip().splitlines():
+            # Use a field-delimiter incompatible with any field value
+            new_nvra = cls.nvra_type(*value.strip().split('#'))
+            if len(new_nvra) != cls.nvra_len:
+                raise ValueError(xcept_fmt % (' '.join(cmd), output))
+            new_name = new_nvra[0]  # documented in class attr. defs. above
+            cls._cache[new_name] = new_nvra
+
+    @classmethod
+    def __getitem__(cls, key):
+        # Don't allow passing arguments to query directly
+        if not key.startswith('-') and key not in cls._cache:
+            cls._store(key)
+        return cls._cache[key]
+
+    @classmethod
+    def fill(cls, method_name=None):
+        """
+        Query and cache details from all packages
+
+        :param method_name: Optional, return cache method with this name
+        """
+        if not cls._alldone:
+            cls._store('-a')  # This is going to take a long while
+            cls._alldone = True
+        if method_name:
+            return getattr(cls._cache, method_name)
+
+    # All of these require a full query
+    __iter__ = classmethod(lambda cls: cls.fill('__iter__')())
+    __len__ = classmethod(lambda cls: cls.fill('__len__')())
+    __contains__ = classmethod(lambda cls, item:
+                               cls.fill('__contains__')(item))
+
+    @classmethod
+    def flush(cls):
+        """Flush cache of RPM NVRAs to force fresh queries on access"""
+        cls._cache = dict()
+        cls._alldone = False
+
+
+class CanonicalDistro(str):
+
+    """
+    Represents a canonical/standardized name for the OS distribution
+
+    :param object: Optional, same as ``str()`` built in.  If non-empty,
+                   take ``str(object`` as the definitive distro. name,
+                   bypassing automatic detection.
+    """
+
+    #: Ordered list of class-method names to use for distro-detection.
+    #: Each takes a dictionary representing the current detection state
+    #: (See as ``matches`` below), and is expected to return
+    #: a string representing the detected distro. name, None, or ''.
+    distros = ("fedora", "rhel", "centos")
+
+    #: Mapping of distro names to lower-case distro-method result strings
+    #: if auto-detection was used (i.e. no initialization arguments passed)
+    matches = None
+
+    def __new__(cls, *args, **dargs):
+        _object = cls._handle_object(*args, **dargs)
+        if _object:
+            return super(CanonicalDistro, cls).__new__(cls, str(_object))
```

Comment:
oh ya, I never updated the unit tests.  Sure I can get them working again...

Purpose: This is preserving the base-class (``str``) API.  It's analogous to any other ``super`` call, except it's polymorphic on the class-object itself.  Having it supports, e.g.:

```python
foobar = CanonicalDistro('foobar')
if foobar.is_enterprise():
    do_something_useful()
```